test/suite/provider.with-multiple-methods.test.js

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2021 Comcast Cable Communications Management, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Provider } from '../../build/sdk/javascript/src/sdk.mjs'
+import { transport } from '../Setup.js'
+
+let providerMethodOneNotificationRegistered = false
+let providerMethodOneRequestDispatched = false
+let providerMethodOneResultSent = false
+let numberOfArgsMethodOne = -1
+let valueOne
+let responseCorrelationIdOne
+
+let providerMethodTwoNotificationRegistered = false
+let providerMethodTwoRequestDispatched = false
+let providerMethodTwoResultSent = false
+let numberOfArgsMethodTwo = -1
+let valueTwo
+let responseCorrelationIdTwo
+
+
+beforeAll( () => {
+
+    class MultiProvider {
+        multiMethodOne(...args) {
```

Comment:
where is the responder arg?
The Promise resolution triggers the ready call, not a response call

Reply:
I think we need to flip that.

Every Provider API needs to respond, but only some will have a ready method (those with the `x-handshake: true` tag extension)

Both of these methods have `x-handshake: false` (although leaving it off would work, too)

Reply:
I don't understand why every provider doesn't just have a handshake.... doesn't hurt anything.

Reply:
Exposing unneeded APIs can cause cause accidental usage. handshake is just for UX focus/visibility race conditions today. Any Providers that don't need that shouldn't get one for no reason. They might call it, and we'd ignore it for now.

But if we ever added another reason for a handshake, they'd be calling it already, and hard to track down side effects might happen.

Reply:
Also, each capability will have it's own provider documentation, so this stuff will be front and center once we have docs. No point in having it there all the time and saying "Ignore this" in half of our Providers.

Reply:
I see it as the opposite, the pattern is exactly the same for ever provider.
```
async my_method(parameters, responder) {
  responder(response)
}
```
Since it is an async method, it is all taken care of for them.
If it doesn't require a handshake, then they just don't do anything, they still just call responder with the response.
If it does require a handshake, they do the exact same thing, just need to `await` in this method until their UI is ready.

I see your implementation as the opposite, they now have to consciously call the handshake callback before they resolve the promise.

Reply:
I think it's easier to explain why I switched it with more realistic examples.

If we're doing a provider method for the RPC `Device.onRequestName`, then it's more natural to _return_ the actual name, since that's how methods work in virtually all programming languages, JS included. Also note that a method like this wouldn't have any parameters:

```javascript
function name() {
  return Promise.resolve(name)
}
```

This looks much like any getter in any system would look like.

The way you're proposing, they'd have to write it like this:

```javascript
function name(parameters, responder) {
  responder(name)
}
```

Doing it the above way requires:

1. adding a parameter to their method that is useless (`parameters`)
2. handling another parameter just to return a value in a strange way that we made up

Now if it's a more UX-centric method, like `PinChallenge.onRequestChallenge` then they're going to need a handshake, but any UX is almost certain to have parameters. Again, having the return value be the result and the handshake in the params:

```javascript
async function challenge(parameters, handshake) {
  let pinPromise = await showSomeUX() // returns a promise chain: 1st for UX ready, 2nd for pin string
  handshake()
  let pin = await pinPromise
  return pin
}
```

If we flip it around and pass the responder:

```javascript
async function challenge(parameters, responder) {
  let pinPromise = await showSomeUX() // returns a promise chain: 1st for UX ready, 2nd for pin string
  handshake()
  // can't do await here, or UX won't show up...
  pinPromise.then( pin => {
    responder(pin)
  })
  return pinPromise
}
```

In the first example, the app partner only has to care about asynchronous details that they choose to implement plus the fact that ALL Firebolt APIs return promises.

In the second, they have to:

1. use only `.then()` syntax (or mix `await` / `then()`)
2. Return "something." Our spec says nothing, but at this point, they'll possibly have their own `Promise` that likely resolves to what we want.
3. Send the result to us with the `responder` (inside a `then()`) at the same time they just returned for the handshake

Yes, it's more consistent from provider to provider, but it's also more complicated.

I'm suggesting that "provider" is a pattern, not an API, so consistency isn't the top concern.

If it's not consistently, it should be simplicity, so I think we should have separate APIs that fit into the "provider" pattern, with details that match that specific API.

Reply:
@kpears201 & I met offline and discussed further.

Summary:

- Agree that provider APIs should return a Promise that resolves to the provided "thing"
- Need to account for a UX "ready" handshake in a forward thinking way that doesn't expose new APIs to apps that aren't fulfilling provider APIs

So the new JavaScript signature will be:

```javascript
async function challenge(parameters, session) {
  let pinPromise = await showSomeUX()
  session.focus()
  return pinPromise
}
```

The `session` parameter looks like:

```typescript
interface ProviderSession {
  correlationId(): string // return the current correlationId for debugging
  focus?(): void // requests focus for the current provider
}
```

The `focus()` method will only be available on providers that have the x-allow-focus tag extension.

This is a good place for extending the API, e.g. we could add `session.requestor()` in the future to expose which app initiated the provider flow.

Reply:
Also, really simple providers can still be very clean:

```javascript
function name() {
  return Promise.resolve('Living Room')
}
```

Reply:
Resolving. Reopen if there's more to discuss